From 9b894553b7459672d6be20be092d5dc23635f3c7 Mon Sep 17 00:00:00 2001
From: Javad Rahimipetroudi <javad.rahimipetroudi@mind.be>
Date: Tue, 10 Dec 2024 17:12:20 +0100
Subject: [PATCH 14/14] drivers-iommu-sun50i-iommu-fix-iommu-on-h616.patch

Signed-off-by: Javad Rahimipetroudi <javad.rahimipetroudi@mind.be>
---
 drivers/iommu/sun50i-iommu.c | 14 ++++++++++----
 1 file changed, 10 insertions(+), 4 deletions(-)

diff --git a/drivers/iommu/sun50i-iommu.c b/drivers/iommu/sun50i-iommu.c
index a3085bed7aaa..32ca263cfff8 100644
--- a/drivers/iommu/sun50i-iommu.c
+++ b/drivers/iommu/sun50i-iommu.c
@@ -452,7 +452,9 @@ static int sun50i_iommu_enable(struct sun50i_iommu *iommu)
 		    IOMMU_TLB_PREFETCH_MASTER_ENABLE(3) |
 		    IOMMU_TLB_PREFETCH_MASTER_ENABLE(4) |
 		    IOMMU_TLB_PREFETCH_MASTER_ENABLE(5));
-	iommu_write(iommu, IOMMU_BYPASS_REG, 0);
+//	iommu_write(iommu, IOMMU_BYPASS_REG, 0);
+	iommu_write(iommu, 0x84, 0);
+	iommu_write(iommu, 0x30, 0);
 	iommu_write(iommu, IOMMU_INT_ENABLE_REG, IOMMU_INT_MASK);
 	iommu_write(iommu, IOMMU_DM_AUT_CTRL_REG(SUN50I_IOMMU_ACI_NONE),
 		    IOMMU_DM_AUT_CTRL_RD_UNAVAIL(SUN50I_IOMMU_ACI_NONE, 0) |
@@ -527,14 +529,17 @@ static void *sun50i_iommu_alloc_page_table(struct sun50i_iommu *iommu,
 	dma_addr_t pt_dma;
 	u32 *page_table;
 
-	page_table = kmem_cache_zalloc(iommu->pt_pool, gfp);
+	//page_table = kmem_cache_zalloc(iommu->pt_pool, gfp);
+	page_table = (u32 *)__get_free_pages(GFP_KERNEL | __GFP_ZERO | GFP_DMA32,
+		get_order(PT_SIZE));
 	if (!page_table)
 		return ERR_PTR(-ENOMEM);
 
 	pt_dma = dma_map_single(iommu->dev, page_table, PT_SIZE, DMA_TO_DEVICE);
 	if (dma_mapping_error(iommu->dev, pt_dma)) {
 		dev_err(iommu->dev, "Couldn't map L2 Page Table\n");
-		kmem_cache_free(iommu->pt_pool, page_table);
+		//kmem_cache_free(iommu->pt_pool, page_table);
+		free_pages((unsigned long)page_table, get_order(PT_SIZE));
 		return ERR_PTR(-ENOMEM);
 	}
 
@@ -550,7 +555,8 @@ static void sun50i_iommu_free_page_table(struct sun50i_iommu *iommu,
 	phys_addr_t pt_phys = virt_to_phys(page_table);
 
 	dma_unmap_single(iommu->dev, pt_phys, PT_SIZE, DMA_TO_DEVICE);
-	kmem_cache_free(iommu->pt_pool, page_table);
+	//kmem_cache_free(iommu->pt_pool, page_table);
+	free_pages((unsigned long)page_table, get_order(PT_SIZE));
 }
 
 static u32 *sun50i_dte_get_page_table(struct sun50i_iommu_domain *sun50i_domain,
-- 
2.47.1

